2-Draft-Accepted/RFCnnnn-PowerShell-as-Login-Shell.md

```diff
@@ -0,0 +1,79 @@
+---
+RFC: RFCnnnn
+Author: Steve Lee
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Console
+Comments Due: June 13, 2019
+Plan to implement: Yes
+---
+
+# Enable PowerShell as Login Shell on POSIX-based systems
+
+POSIX shells on Unix-based systems use a `-l` switch to start the shell as a login shell.
+This means the shell is not being started from an existing configured environment and
+needs to execute `/etc/profile` to populate environment variables along with other
+environment changes.
+
+## Motivation
+
+    As a PowerShell user,
+    I can use PowerShell as my login shell on Unix-based systems,
+    so that I can use PowerShell everywhere.
+
+## Specification
+
+Many existing tools expect `-l` to be accepted by a shell and to act as a login
+shell.
+
+`-l` (expanded form is `-LoadProfile`) will explicitly have pwsh load the PowerShell
+profile.
+This is effectively a no-op as if this switch is not specified, pwsh will load
+the PowerShell profile and only doesn't load it if `-noprofile` is specified.
+
+This will allow tools that expect `-l` to be accepted to work.
+There is no additional work to process `/etc/profile` when `-l` is used.
+
+For cases where you do need `/etc/profile` to be processed,
+such as using pwsh as your default shell,
+the proposal is to include a Bourne shell script that can be specified as the
+shell:
+
+```sh
+#!/bin/sh -l
+exec /usr/local/bin/pwsh "$@"
+```
+
+>[!NOTE]
+> The actual path would depend on the operation system and whether PowerShell
+> is a stable or preview release.
+> For Linux, the path would be `/usr/bin`.
+> For Snap, the path would be `/snap/bin`.
+> For macOS, the path would be `/usr/local/bin`.
+> Executable would be `pwsh-preview` for preview releases.
+
+This script will be called `pwsh-login` and should be used whenever you require
+a specific login shell.
+
+## Alternate Proposals and Considerations
+
+### Process /etc/profile using sh and copy the env vars
+
+This proposal would be that if `-l` is specified, pwsh would run `/bin/sh -l -c export`
+which would create a new shell process that does all the processing needed for
+a login shell and export the environment variables that would be exported.
+Additional code is needed to take those environment variables and copy them to
+the parent pwsh process.
+
+The downsides to this approach is additional code to maintain,
+but more importantly not getting a complete login shell environment as things
+like ulimit and umask would not be inherited into pwsh.
+
+### pwsh to start `/bin/sh -l -c "exec pwsh"`
```

Comment:
I'm not entirely opposed to this, since in a greenfield we could implement `-l` before anything else in startup and get it to `execv` the `pwsh-login` script... But not sure how feasible that really is

Reply:
Some OSes have limit on parameter account.

Reply:
The problem with `-l` doing any `exec` is that it results in incurring twice the startup perf penalty as `pwsh` is effectively started twice.

Reply:
> twice the startup perf penalty as pwsh is effectively started twice

I'm imagining a pretty different path on the first startup; no modules, no runspaces, no engine, no filesystem skullduggery, just a beeline for `exec`. Do we know what the difference is between our startup time and a .NET Core `"Hello World"`?

In short, I don't imagine *twice* the startup time. Perhaps an added 10%?

Reply:
You still have to load the main `pwsh` binary, which is much bigger than a .NET Core `"Hello World!".

Reply:
Well I'm not sure binary size is the limiting factor on startup time. My suspicion is that it's the size of the JIT-ed code path.

We need some data here I think. I discussed this with @daxian-dbw and I'm going to look into a prototype of my proposal.

Reply:
@rjmholt if you have some perf results, we can revisit this.  We have an opportunity to change this if needed before the last Preview to get feedback.

Reply:
Tried it here: https://github.com/rjmholt/PowerShell/commit/eb6c5f718e90052b633ca9f602a2d736ae4be05c

Looks like an 8.5% perf hit:

```powershell
impo build.psm1; start-psbuild -clean
$pwsh = get-psoutput

$times = 1000
$withoutLogin = measure-command { foreach ($x in 1..$times) { & $pwsh  -NoProfile -NoLogo -Command 'exit' } }
$withLogin = measure-command { foreach ($x in 1..$times) { & $pwsh  -NoProfile -Login -NoLogo -Command 'exit' } }

$withLogin / $withoutLogin - 1     # 0.085355849110186
```

Reply:
And how does performance of the
```sh
#!/bin/sh -l
exec /usr/local/bin/pwsh "$@"
```
script compare to this? (I expect it will be much faster)

Reply:
A 8.5% perf hit seems acceptable for a login shell as it's doing more (like processing the system profile to populate env vars, etc...).  I think I'll update the RFC to propose @rjmholt's proposal as the preferred one.